fendermint/testing/materializer/src/lib.rs

```diff
@@ -1,6 +1,257 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "'{}'", self.0)
+    }
+}
+
+impl From<&str> for ResourceId {
+    fn from(value: &str) -> Self {
+        Self::from(value.to_string())
+    }
+}
+
+impl From<String> for ResourceId {
+    fn from(value: String) -> Self {
+        Self(value.replace('/', "_"))
+    }
+}
+
+impl From<&ResourceId> for ResourceId {
+    fn from(value: &Self) -> Self {
+        value.clone()
+    }
+}
+
+/// A human readable name for a testnet.
+pub type TestnetId = ResourceId;
+
+/// A human readable name for an account.
+pub type AccountId = ResourceId;
+
+/// A human readable name for a subnet.
+pub type SubnetId = ResourceId;
+
+/// A human readable name for a node.
+pub type NodeId = ResourceId;
+
+/// A human readable name for a relayer.
+pub type RelayerId = ResourceId;
+
+/// The name of a resource consists of its ID and all the IDs of its ancestors
+/// concatenated into a URL-like path.
+///
+/// See <https://cloud.google.com/apis/design/resource_names>
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ResourceName(PathBuf);
+
+impl ResourceName {
+    fn join(&self, s: &str) -> Self {
+        Self(self.0.join(s))
+    }
+
+    fn join_id(&self, id: &ResourceId) -> Self {
+        self.join(&id.0)
+    }
+
+    pub fn is_prefix_of(&self, other: &ResourceName) -> bool {
+        other.0.starts_with(&self.0)
+    }
+}
+
+impl From<&str> for ResourceName {
+    fn from(value: &str) -> Self {
+        Self(PathBuf::from(value))
+    }
+}
+
+impl Display for ResourceName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "'{}'", self.0.to_string_lossy())
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TestnetName(ResourceName);
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AccountName(ResourceName);
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SubnetName(ResourceName);
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct NodeName(ResourceName);
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RelayerName(ResourceName);
+
+impl TestnetName {
+    pub fn new<T: Into<TestnetId>>(id: T) -> Self {
+        Self(ResourceName::from("/testnets").join_id(&id.into()))
+    }
+
+    pub fn account<T: Into<AccountId>>(&self, id: T) -> AccountName {
+        AccountName(self.0.join("accounts").join_id(&id.into()))
+    }
+
+    pub fn root(&self) -> SubnetName {
+        SubnetName(self.0.join("root"))
+    }
+}
+
+impl NodeName {
+    pub fn is_in_subnet(&self, subnet_name: &SubnetName) -> bool {
+        subnet_name.0.is_prefix_of(&self.0)
+    }
+}
+
+impl SubnetName {
+    pub fn subnet<T: Into<SubnetId>>(&self, id: T) -> Self {
```

Comment:
nice, this implementation is pretty clever!

Reply:
I wanted to make it fluent as much as I could, without a lot of `.into()`, so something like `.subnet("foo")` should work in tests.